r_gl.cpp

```diff
@@ -1079,10 +1080,24 @@ void GLRender::RenderConsole(Console* console, CFont* font) {
         console->m_blinkTimer = 0;
     }
 
+    // draw bottom scroll indicator
+    if (isScrollable && console->ScrollPos() != 0) {
+        R_DrawText(std::string(maxChars, '^'), textMargin, logY + lineHeight * 0.7f, COORD_MODE_ABS);
+    }
+
     // draw log lines
     for (int i = 0; i < maxLines; i++) {
+        int lineOffset = isScrollable ? console->ScrollPos() + i : i;
         std::string line;
-        if (!console->m_lineBuffer.Get(i, &line)) break;
+        if (!console->m_lineBuffer.Get(lineOffset, &line)) {
+            if (isScrollable) {
+                std::string msg = " END OF LOG ";
+                int chars = (maxChars - msg.length()) / 2;
+                msg = std::string(chars, '-') + msg;
+                R_DrawText(msg + std::string(maxChars - msg.length(), '-'), textMargin, logY, COORD_MODE_ABS);
```

Comment:
Note that a new string object is allocated here all the time. It is expensive. We do this quite often throughout the codebase not just here. Just something to keep in mind ;)